async_task.go

```diff
@@ -88,12 +90,12 @@ func (t *TaskStatus) Wait(ctx context.Context) (interface{}, error) {
 	case <-ch:
 		return t.result, t.err
 	case <-ctx.Done():
-		t.finish(StateCanceled, nil, ErrTimeout)
-		return t.result, t.err
+		return nil, ErrWaitTimeout
 	}
 }
 
 // WaitWithTimeout block current thread/routine until task finished or failed, or exceed the duration specified.
+// timeout only stop waiting, taks will remain running.
 func (t *TaskStatus) WaitWithTimeout(ctx context.Context, timeout time.Duration) (interface{}, error) {
```

Comment:
why wouldn't you rely on the context to setup the timeout? 
`context.WithTimeout(context.Background(), 20*time.Second)`

Reply:
**suggestion** : leave it to the caller to setup a timeout on the context passed in.

Reply:
the caller can definitely call Wait() with a context already setup.

or caller can invoke this shortcut, which make code easy to read.

```
ctxForWait, cancelFunc := context.WithTimeout(ctx, 2*time.Second)
defer cancelFunc() // or you get a warning for ignore this cancelFunc
tsk.Wait(ctxForWait)
```
vs 
```
context.WaitWithTimeout(ctx, 2*time.Second)
```